File: simulations/nonlinear.py
"""
Reproduce Figures 19 and 20  (nonlinear confounding effects).

alpha controls nonlinearity in H -> X:   X = nlX(alpha, H @ Gamma) + E
beta  controls nonlinearity in H -> Y:   Y = f(X) + nlY(beta,  H @ psi) + e

where nlX(a, t) = (1-a)*t + a*|t|  interpolates linearly between t and |t|.

Results are shown as 7×7 heatmaps of average MSE ratios over the (alpha, beta) grid.

Equivalent to R's SimulationScripts/Nonlinear.R.

Usage
-----
python nonlinear.py              # run simulations + plot
python nonlinear.py --plot-only  # load saved results and plot
"""

from __future__ import annotations
import argparse
import sys
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from hdam import fit_deconfounded_hdam, fit_hdam_with_est_factors, estimate_function
from _sim_utils import (
    f_true,
    run_parallel, save_results, load_results,
)

# ---------------------------------------------------------------------------
# Fixed parameters
# ---------------------------------------------------------------------------

Q = 5
N = 400
P = 500
N_REP = 100

AL_VEC  = list(np.linspace(0, 1, 7))   # alpha values
BET_VEC = list(np.linspace(0, 1, 7))   # beta  values

CI_VEC = [
    ("equal",      False),
    ("decreasing", True),
]


# ---------------------------------------------------------------------------
# Nonlinearity helpers  (equivalent to R's nlX / nlY)
# ---------------------------------------------------------------------------

def nlX(al: float, t: np.ndarray) -> np.ndarray:
    """Linear interpolation between t and |t|."""
    return (1 - al) * t + al * np.abs(t)


def nlY(bet: float, t: np.ndarray) -> np.ndarray:
    """Linear interpolation between t and |t|."""
    return (1 - bet) * t + bet * np.abs(t)


# ---------------------------------------------------------------------------
# Single simulation run
# ---------------------------------------------------------------------------

def one_sim(al, bet, q, n, p, seed_val, decreasing_confounding=False):
    rng = np.random.default_rng(seed_val)

    Gamma = rng.uniform(-1, 1, size=(q, p))
    if decreasing_confounding:
        Gamma = (1.0 / np.arange(1, q + 1))[:, None] * Gamma

    psi = rng.uniform(0, 2, size=q)
    H   = rng.standard_normal((n, q))
    E   = rng.standard_normal((n, p))
    e   = 0.5 * rng.standard_normal(n)

    X = nlX(al, H @ Gamma) + E
    Y = f_true(X) + nlY(bet, H @ psi) + e

    kw = dict(n_K=5, cv_method="1se", cv_k=5, n_lambda1=10, n_lambda2=25)
    lres_trim    = fit_deconfounded_hdam(Y, X.copy(), meth="trim", **kw)
    lres_none    = fit_deconfounded_hdam(Y, X.copy(), meth="none", **kw)
    lres_est_fac = fit_hdam_with_est_factors(Y, X.copy(), **kw)

    n_test = 1000
    H_test = rng.standard_normal((n_test, q))
    E_test = rng.standard_normal((n_test, p))
    X_test = nlX(al, H_test @ Gamma) + E_test
    f_test = f_true(X_test)

    return {
        "MSE": [
            float(np.mean((f_test - estimate_function(X_test, lres_trim)) ** 2)),
            float(np.mean((f_test - estimate_function(X_test, lres_none)) ** 2)),
            float(np.mean((f_test - estimate_function(X_test, lres_est_fac)) ** 2)),
        ],
        "active": [
            len(lres_trim["active"]),
            len(lres_none["active"]),
            len(lres_est_fac["active"]),
        ],
    }


# ---------------------------------------------------------------------------
# Run simulations
# ---------------------------------------------------------------------------

def run_simulations(out_dir: Path, n_rep: int, n_cores: int) -> None:
    out_dir.mkdir(parents=True, exist_ok=True)

    rng_master = np.random.default_rng(1432)
    seed_vec = rng_master.integers(1, 2_100_000_000, size=n_rep).tolist()

    for i, al in enumerate(AL_VEC, start=1):
        for j, bet in enumerate(BET_VEC, start=1):
            for ci_tag, dec in CI_VEC:
                fname = out_dir / f"{ci_tag}CI_al_{i}_bet_{j}.pkl"
                if fname.exists():
                    print(f"[skip] {fname.name}")
                    continue
                print(f"[run ] al={al:.3f}, bet={bet:.3f}, {ci_tag} CI ...")
                args = [(al, bet, Q, N, P, sv, dec) for sv in seed_vec]
                results = run_parallel(one_sim, args, n_cores)
                save_results(results, fname)
                print(f"       saved {fname.name}")


# ---------------------------------------------------------------------------
# Compute average MSE matrices
# ---------------------------------------------------------------------------

def _avg_mse_matrix(out_dir: Path, n_rep: int, ci_tag: str, meth_idx: int) -> np.ndarray:
    """Return 7x7 matrix of average MSE[meth_idx] for the given CI setting."""
    mat = np.full((len(AL_VEC), len(BET_VEC)), np.nan)
    for i, _ in enumerate(AL_VEC):
        for j, _ in enumerate(BET_VEC):
            fname = out_dir / f"{ci_tag}CI_al_{i+1}_bet_{j+1}.pkl"
            if not fname.exists():
                continue
            results = load_results(fname)
            mat[i, j] = np.mean([r["MSE"][meth_idx] for r in results[:n_rep]])
    return mat


# ---------------------------------------------------------------------------
# Plots (Figures 19 and 20)
# ---------------------------------------------------------------------------

def _heatmap(ax, ratio_mat, al_vec, bet_vec, title: str) -> None:
    """
    Draw a heatmap of ratio_mat with text annotations.
    Uses a reversed magma-like colormap (dark = high ratio).
    """
    im = ax.imshow(
        ratio_mat,
        origin="lower",
        aspect="auto",
        cmap="magma_r",
        vmin=ratio_mat[np.isfinite(ratio_mat)].min(),
        vmax=ratio_mat[np.isfinite(ratio_mat)].max(),
    )
    ax.set_xticks(range(len(bet_vec)))
    ax.set_yticks(range(len(al_vec)))
    ax.set_xticklabels([f"{v:.2f}" for v in bet_vec], fontsize=8)
    ax.set_yticklabels([f"{v:.2f}" for v in al_vec],  fontsize=8)
    ax.set_xlabel(r"$\beta$", fontsize=11)
    ax.set_ylabel(r"$\alpha$", fontsize=11)
    ax.set_title(title, fontsize=10)

    for i in range(ratio_mat.shape[0]):
        for j in range(ratio_mat.shape[1]):
            val = ratio_mat[i, j]
            if np.isfinite(val):
                # white text on dark cells, black on light cells
                color = "white" if val >= 0.995 else "black"
                ax.text(j, i, f"{val:.2f}", ha="center", va="center",
                        fontsize=7, color=color)


def make_plots(out_dir: Path, plot_dir: Path, n_rep: int) -> None:
    plot_dir.mkdir(parents=True, exist_ok=True)

    # meth indices: 0=trim(deconfounded), 1=none(naive), 2=estFac
    meth_names = ["deconfounded", "naive", "estimated factors"]

    ci_labels = {
        "equal":      "equal confounding influence",
        "decreasing": "decreasing confounding influence",
    }

    for ci_tag, ci_lbl in ci_labels.items():
        mat_dec  = _avg_mse_matrix(out_dir, n_rep, ci_tag, 0)  # deconfounded
        mat_naive= _avg_mse_matrix(out_dir, n_rep, ci_tag, 1)  # naive
        mat_ef   = _avg_mse_matrix(out_dir, n_rep, ci_tag, 2)  # est. factors

        if np.all(np.isnan(mat_dec)):
            print(f"No results for {ci_tag}; skipping plots.")
            continue

        ratio_dec_vs_naive = mat_dec / mat_naive
        ratio_dec_vs_ef    = mat_dec / mat_ef

        fig, axes = plt.subplots(1, 2, figsize=(12, 5))
        fig.suptitle(
            f"Ratio of average MSE \u2013 {ci_lbl}", fontsize=12,
        )

        _heatmap(axes[0], ratio_dec_vs_naive, AL_VEC, BET_VEC,
                 "deconfounded vs. naive")
        _heatmap(axes[1], ratio_dec_vs_ef,    AL_VEC, BET_VEC,
                 "deconfounded vs. estimated factors")

        fig.tight_layout()
        out = plot_dir / f"Nonlinear_{ci_tag}CI.pdf"
        fig.savefig(out)
        plt.close(fig)
        print(f"saved {out.name}")


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--plot-only", action="store_true")
    parser.add_argument("--n-rep",   type=int, default=N_REP)
    parser.add_argument("--n-cores", type=int, default=4)
    args = parser.parse_args()

    out_dir  = Path("SimulationResults/Nonlinear")
    plot_dir = Path("PlotResults/Nonlinear")

    if not args.plot_only:
        run_simulations(out_dir, args.n_rep, args.n_cores)

    make_plots(out_dir, plot_dir, args.n_rep)

File: simulations/Nonlinear.R
## Code to reproduce Figures 19 and 20
# Nonlinear confounding effects

# Import the functions needed
source("../r/FitDeconfoundedHDAM.R")
source("../r/AnalyzeFittedHDAM.R")
source("../r/EstimationFactors.R")


library("parallel")
library(ggplot2)

f1 <- function(x){-sin(2*x)}
f2 <- function(x){2-2*tanh(x+0.5)}
f3 <- function(x){x}
f4 <- function(x){4/(exp(x)+exp(-x))}
f <- function(X){f1(X[,1])+f2(X[,2])+f3(X[,3])+f4(X[,4])}

# function to induce nonlinearity in H->X
# linear interpolation between t and abs(t)
nlX <- function(al, t){(1-al)*t+al*abs(t)}

# function to induce nonlinearity in H->Y
# linear interpolation between t and abs(t)
nlY <- function(bet, t){(1-bet)*t+bet*abs(t)}

# fix n = 400, p = 500
q <- 5
n <-  400
p <- 500


one.sim <- function(al, bet, seed.val, decreasing.confounding.influence = FALSE){
  set.seed(seed.val)
  # generate data
  Gamma <- matrix(runif(q*p, min = -1, max = 1), nrow=q)
  if(decreasing.confounding.influence){
    gam.weight <- 1/(1:q)
    Gamma <- gam.weight * Gamma
  }
  psi <- runif(q, min = 0, max = 2)
  H <- matrix(rnorm(n*q), nrow = n)
  E <- matrix(rnorm(n*p), nrow = n)
  e <- 0.5*rnorm(n)
  X <- nlX(al, H %*% Gamma) + E
  Y <- f(X) + nlY(bet, H %*% psi) + e
  lres.trim <- FitDeconfoundedHDAM(Y, X, n.K = 5, meth = "trim", cv.method = "1se", cv.k = 5, n.lambda1 = 10, n.lambda2 = 25)
  lres.none <- FitDeconfoundedHDAM(Y, X, n.K = 5, meth = "none", cv.method = "1se", cv.k = 5, n.lambda1 = 10, n.lambda2 = 25)
  lres.estFac <- FitHDAM.withEstFactors(Y, X, n.K = 5, cv.method = "1se", cv.k = 5, n.lambda1 = 10, n.lambda2 = 25)
  ntest <- 1000
  Htest <- matrix(rnorm(ntest * q), nrow = ntest)
  Etest <- matrix(rnorm(ntest * p), nrow = ntest)
  Xtest <- nlX(al, Htest %*% Gamma) + Etest
  fhat.trim <- estimate.function(Xtest, lres.trim)
  fhat.none <- estimate.function(Xtest, lres.none)
  fhat.estFac <- estimate.function(Xtest, lres.estFac)
  fXtest <- f(Xtest)
  mat.return <- rbind(c(mean((fXtest-fhat.trim)^2), mean((fXtest-fhat.none)^2), mean((fXtest-fhat.estFac)^2)),
                      c(length(lres.trim$active), length(lres.none$active), length(lres.estFac$active)))
  rownames(mat.return) <- c("MSE", "Active")
  colnames(mat.return) <- c("trim", "none", "estFac")
  return(mat.return)
}


nrep <- 100
use.cores <- 50



ta <- Sys.time()
al.vec <- seq(0, 1, length.out=7)
bet.vec <- seq(0, 1, length.out=7)


set.seed(1432)
smax <- 2100000000
seed.vec <- sample(1:smax, nrep)

for (i in 1:length(al.vec)){
  al <- al.vec[i]
  for (j in 1:length(bet.vec)){
    bet <- bet.vec[j]
    # with equal confounding influence
    fun = function(seed.val){return(one.sim(al = al, bet = bet, seed.val = seed.val, decreasing.confounding.influence = FALSE))}
    lres <- mclapply(seed.vec, fun, mc.cores = use.cores)
    filename <- paste("SimulationResults/NL_2024_08_13/equalCI_al_", i,"_bet_", j, ".RData", sep = "")
    save(lres, file = filename)
    # with decreasing confounding influence
    fun = function(seed.val){return(one.sim(al = al, bet = bet, seed.val = seed.val, decreasing.confounding.influence = TRUE))}
    lres <- mclapply(seed.vec, fun, mc.cores = use.cores)
    filename <- paste("SimulationResults/NL_2024_08_13/decreasingCI_al_", i,"_bet_", j, ".RData", sep = "")
    save(lres, file = filename)
  }
}
te <- Sys.time()
time.needed <- te-ta
print(paste("Time needed: ", time.needed))



calc.average <- function(quantity, CI, meth, i, j){
  filename <- paste("SimulationResults/NL_2024_08_13/", CI, "CI_al_", i,"_bet_", j, ".RData", sep = "")
  load(filename)
  return(mean(unlist(lapply(lres, function(mat){mat[quantity, meth]}))))
}

quantity.vec <- c("MSE", "Active")
CI.vec <- c("equal", "decreasing")
meth.vec <- c("trim", "none", "estFac")
for(quantity in quantity.vec){
  for(CI in CI.vec){
    for(meth in meth.vec){
      varname <- paste("mat_", quantity, "_", CI, "CI_", meth, sep = "")
      assign(varname, outer(1:7, 1:7, FUN = Vectorize(function(i,j){calc.average(quantity, CI, meth, i, j)})))
    }
  }
}


meth.name.vec <- c("deconfounded", "naive", "estimated factors")
library(viridisLite)
plot_ratios <- function(i, j, CI){
  MSE_meth1 <- get(paste("mat_MSE_", CI, "CI_", meth.vec[i], sep = ""))
  MSE_meth2 <- get(paste("mat_MSE_", CI, "CI_", meth.vec[j], sep = ""))
  mat_ratio <- MSE_meth1/MSE_meth2
  main_text <- paste( meth.name.vec[i], " vs. ", meth.name.vec[j], sep = "")
  image(al.vec, bet.vec, mat_ratio, axes = FALSE, col = magma(100, direction = -1), xlab = expression(alpha), ylab = expression(beta),
       main = main_text)
  axis(1, al.vec, round(al.vec, digits = 2))
  axis(2, bet.vec, round(bet.vec, digits = 2))
  for (i in 1:length(al.vec)){
    for (j in 1:length(bet.vec)){
      if(mat_ratio[i,j] < 0.995){
        text(al.vec[i], bet.vec[j], round(mat_ratio[i,j], 2), cex = 1.3)
      }
      else{
        text(al.vec[i], bet.vec[j], round(mat_ratio[i,j], 2), cex = 1.3, col = "white")
      }
    }
  }
}


par(mfrow = c(1,2), oma = c(0,0, 1, 0))
plot_ratios(1, 2, "equal")
plot_ratios(1, 3, "equal")

mtext("Ratio of average MSE \u2013 equal confounding influence", outer = TRUE, cex = 1.5, line = -1)

par(mfrow = c(1,2), oma = c(0,0, 1, 0))
plot_ratios(1, 2, "decreasing")
plot_ratios(1, 3, "decreasing")

mtext("Ratio of average MSE \u2013 decreasing confounding influence", outer = TRUE, cex = 1.5, line = -1)


